tls/extensions/s2n_server_psk.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <sys/param.h>
+#include <stdint.h>
+
+#include "tls/s2n_tls.h"
+#include "tls/extensions/s2n_server_psk.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_bitmap.h"
+
+static bool s2n_server_psk_should_send(struct s2n_connection *conn);
+static int s2n_server_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out);
+static int s2n_server_psk_recv(struct s2n_connection *conn, struct s2n_stuffer *extension);
+
+const s2n_extension_type s2n_server_psk_extension = {
+    .iana_value = TLS_EXTENSION_PRE_SHARED_KEY,
+    .is_response = true,
+    .send = s2n_server_psk_send,
+    .recv = s2n_server_psk_recv,
+    .should_send = s2n_server_psk_should_send,
+    .if_missing = s2n_extension_noop_if_missing,
+};
+
+static bool s2n_server_psk_should_send(struct s2n_connection *conn)
+{
+    /* Only send a server pre_shared_key extension if a chosen PSK is set on the connection */
+    return conn && s2n_connection_get_protocol_version(conn) >= S2N_TLS13
+            && conn->psk_params.chosen_psk;
+}
+
+static int s2n_server_psk_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    notnull_check(conn);
+
+    /* Send the index of the chosen PSK that is stored on the connection. */
+    GUARD(s2n_stuffer_write_uint16(out, conn->psk_params.chosen_psk_wire_index));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_server_psk_recv(struct s2n_connection *conn, struct s2n_stuffer *extension)
+{
+    notnull_check(conn);
+
+    if (s2n_connection_get_protocol_version(conn) < S2N_TLS13) {
+        return S2N_SUCCESS;
+    }
+
+    /* A key_share extension must have been received in order to use a pre-shared key
+     * in (EC)DHE key exchange mode.
+     */
+    s2n_extension_type_id key_share_ext_id;
+    GUARD(s2n_extension_supported_iana_value_to_id(TLS_EXTENSION_KEY_SHARE, &key_share_ext_id));
+    ENSURE_POSIX(S2N_CBIT_TEST(conn->extension_requests_received, key_share_ext_id), S2N_ERR_MISSING_EXTENSION);
```

Comment:
I'm not sure if this is an issue or not, but if we eventually support psk without ECDHE key exchange, the server would not send a keyshare ext. and this line would error. Is it better behavior to exit just here? I think at this point there's still a possibility of a complete handshake.

Reply:
We do not support PSK without ECDHE key exchange mode and the contract requires that the key_share extension MUST have been received to use PSK, which is ensured/enforced in this check. Exiting without an error would mean that we are have an undefined behavior, which is not the case.

Reply:
Currently, we do not support PSK without ECDHE key exchange mode and the contract requires that the key_share extension MUST have been received to use PSK, which is ensured/enforced in this check. If in the future we do support PSK without ECDHE key exchange mode, this logic will have to be tweaked. Exiting without an error currently would mean that we are have an undefined behavior, which is not the case as receiving a psk extension without key_share_extension is defined and is an error case.